lib/models/addon.js

```diff
@@ -1445,4 +1442,112 @@ Addon.lookup = function(addon) {
+    var self = this;
+    var methodsToValidate = methodsForTreeType(treeType);
+
+    // determine if treeFor* (or other methods for tree type) overrides for the given tree
+    var anyMethodsModified = methodsToValidate.some(function(methodName) {
+      return self[methodName] !== addonProto[methodName];
+    });
+
+    if (anyMethodsModified) {
+      return null; // uncacheable
+    }
+
+    // determine if treeMethodsFor overrides for given tree
+    if (self.treeForMethods[treeType] !== DEFAULT_TREE_FOR_METHODS[treeType]) {
+      return null; // uncacheable
+    }
+
+    // compute cache key
+    var cacheKey = calculateCacheKeyForTree(treeType, self);
+
+    return cacheKey; // profit?
+  };
+}
+
+var Addon = CoreObject.extend(addonProto);
+
+/**
+  Returns the absolute path for a given addon
+
+  @private
+  @method resolvePath
+  @param {String} addon Addon name
+  @return {String} Absolute addon path
+*/
+Addon.resolvePath = function(addon) {
+  var addonMain = addon.pkg['ember-addon-main'];
+
+  if (addonMain) {
+    this.ui && this.ui.writeDeprecateLine(addon.pkg.name + ' is using the deprecated ember-addon-main definition. It should be updated to {\'ember-addon\': {\'main\': \'' + addon.pkg['ember-addon-main'] + '\'}}');
+  } else {
+    addonMain = (addon.pkg['ember-addon'] && addon.pkg['ember-addon'].main) || addon.pkg['main'] || 'index.js';
+  }
+
+  // Resolve will fail unless it has an extension
+  if (!path.extname(addonMain)) {
+    addonMain += '.js';
+  }
+
+  return path.resolve(addon.path, addonMain);
+};
+
+/**
+  Returns the addon class for a given addon name.
+  If the Addon exports a function, that function is used
+  as constructor. If an Object is exported, a subclass of
+  `Addon` is returned with the exported hash merged into it.
+
+  @private
+  @static
+  @method lookup
+  @param {String} addon Addon name
+  @return {Addon} Addon class
+*/
+Addon.lookup = function(addon) {
+  var Constructor, addonModule, modulePath, moduleDir;
+  var start = Date.now();
+
+  modulePath = Addon.resolvePath(addon);
+  moduleDir  = path.dirname(modulePath);
+
+  if (existsSync(modulePath)) {
+    addonModule = require(modulePath);
+
+    if (typeof addonModule === 'function') {
+      Constructor = addonModule;
+      Constructor.prototype.root = Constructor.prototype.root || moduleDir;
+      Constructor.prototype.pkg  = Constructor.prototype.pkg || addon.pkg;
+    } else {
+      Constructor = Addon.extend(assign({
+        root: moduleDir,
+        pkg: addon.pkg
+      }, addonModule));
+    }
+  }
+
+  if (!Constructor) {
+    throw new SilentError('The `' + addon.pkg.name + '` addon could not be found at `' + addon.path + '`.');
+  }
+
+  var lookupDuration = Date.now() - start;
+
+  Constructor._meta_ = {
+    modulePath: modulePath,
+    lookupDuration: lookupDuration,
+    initializeIn: 0
+  };
+
+  return Constructor;
+};
 module.exports = Addon;
+module.exports._resetTreeCache = _resetTreeCache;
```

Comment:
`broccoli-persistent-filter` is all about caching across restarts, however the cache that we are working with here is not persistent across restarts.  The cache that this PR implements is strictly to ensure that the same addon (version/name/dependencies/etc) at multiple levels of depth (in the addon hierarchy) return the same tree.

_tldr;_ This cache cannot interact with things using on disk caches (since it doesn't survive a restart).